src/core/components/block.js

```diff
@@ -11,43 +12,32 @@ module.exports = function block (self) {
       cid = cleanCid(cid)
       self._blockService.get(cid, callback)
     },
-    put: (block, cid, callback) => {
-      if (typeof cid === 'function') {
-        // legacy (without CID)
-        callback = cid
-        cid = undefined
-      }
-
+    put: (block, callback) => {
       if (Array.isArray(block)) {
         return callback(new Error('Array is not supported'))
       }
 
-      if (Buffer.isBuffer(block)) {
-        block = new Block(block)
-      }
-
       waterfall([
         (cb) => {
-          if (cid) {
-            return cb(null, cid)
+          if (Block.isBlock(block)) {
+            return cb(null, block)
           }
 
-          block.key('sha2-256', (err, key) => {
+          multihashing(block, 'sha2-256', (err, multihash) => {
             if (err) {
               return cb(err)
             }
 
-            cb(null, new CID(key))
+            cb(null, new Block(block, new CID(multihash)))
           })
         },
-        (cid, cb) => self._blockService.put({block: block, cid: cid}, cb)
-      ], (err) => {
-        if (err) {
-          return callback(err)
-        }
-
-        callback(null, block)
-      })
+        (block, cb) => self._blockService.put(block, (err) => {
+          if (err) {
+            return cb(err)
+          }
+          cb(null, block)
+        })
+      ], callback)
```

Comment:
This has changed the API, block needs to be able to:

```
» jsipfs block put --help
jsipfs block put [block]

Options:
  --help        Show help                                              [boolean]
  --format, -f  cid format for blocks to be created with.    [default: "dag-pb"]
  --mhtype      multihash hash function                    [default: "sha2-256"]
  --mhlen       multihash hash length
```

Before we passed that through CID, if you remove the CID from equation and always calculate the multihash of sha256, then you can't block put eth data anymore.